api/plugin/vproto/api.proto

```diff
@@ -0,0 +1,317 @@
+  //
+  // The response is empty. It is not possible for a plugin to update
+  // containers in response to this request.
+  rpc RunPodSandbox(RunPodSandboxRequest) returns (google.protobuf.Empty);
+
+  // StopPodSandbox relays the corresponding CRI event to the plugin.
+  //
+  // The response is empty. It is not possible for a plugin to update
+  // containers in response to this request. All containers in the pod
+  // will be explicitly stopped before the pod is stopped.
+  rpc StopPodSandbox(StopPodSandboxRequest) returns (google.protobuf.Empty);
+
+  // RemovePodSandbox relays the corresponding CRI event to the plugin.
+  //
+  // The response is empty. It is not possible for a plugin to update
+  // containers in response to this request. All containers in the pod
+  // will be explicitly stopped and removed before removing the pod.
+  rpc RemovePodSandbox(RemovePodSandboxRequest) returns (google.protobuf.Empty);
+
+  // CreateContainer relays the corresponding CRI event to the plugin.
+  //
+  // The plugin responds with any modifications to the container being created
+  // and updates to other containers.
+  rpc CreateContainer(CreateContainerRequest) returns (CreateContainerResponse);
+
+  // StartContainer relays the corresponding CRI event to the plugin.
+  //
+  // The response is empty. It is not possible for a plugin to update
+  // containers in response to this request.
+  rpc StartContainer(StartContainerRequest) returns (google.protobuf.Empty);
+
+  // UpdateContainer relays the corresponding UpdateContainerResources event
+  // to the plugin.
+  //
+  // The plugin responds with any updates to the container being modified
+  // and to other containers.
+  rpc UpdateContainer(UpdateContainerRequest) returns (UpdateContainerResponse);
+
+  // StopContainer relays the corresponding CRI event to the plugin.
+  //
+  // The plugin responds with any updates to remaining containers in the
+  // system.
+  rpc StopContainer(StopContainerRequest) returns (StopContainerResponse);
+
+  // RemoveContainer relays the corresponding CRI event to the plugin.
+  //
+  // The response is empty. It is not possible for a plugin to update
+  // containers in response to this request. All containers will be explicitly
+  // stopped before removal.
+  rpc RemoveContainer(RemoveContainerRequest) returns (google.protobuf.Empty);
+}
+
+// Event is used by plugins to subscribe for pod and container lifecycle
+// events.
+enum Event {
+  // Hook into RunPodSandbox.
+  RUN_POD_SANDBOX = 0;
+  // Hook into StopPodSandbox.
+  STOP_POD_SANDBOX = 1;
+  // Hook into RemovePodSandbox.
+  REMOVE_POD_SANDBOX = 2;
+  // Hook into CreateContainer.
+  CREATE_CONTAINER = 3;
+  // Hook into StartContainer.
+  START_CONTAINER = 4;
+  // Hook into UpdateContainer.
+  UPDATE_CONTAINER = 5;
+  // Hook into StopContainer.
+  STOP_CONTAINER = 6;
+  // Hook into RemoveContainer.
+  REMOVE_CONTAINER = 7;
+  // Hook into all events.
+  ALL = 8;
```

Comment:
[low priority] This is more food-for-thought rather than a change request. But possibly something to consider for future extensibility.

At some point we might want to extend `Events` with crazy things like
- `HEARTBEAT` maybe with a couple of intervals - this would enable implementing rebalancing workloads, periodic memory arrangements etc.
- `PLUGIN_EVENT` that could be dispatched by a plugin and captured by another plugin and/or runtime. For instance, a plugin could have discovered new memory device being attached to the node over the fabric.

Regarding NRI-plugin backwards compatibility, these examples would mean:
1. `ALL = 8` event is problematic, because a plugin that subscribes to every event existing now cannot know beforehand what events new NRI interface will actually serve when it starts the plugin. Do you think we could be fine with a list of explicit events to subscribe to?
2. At some point some events could have parameters, like plugin's wish for heartbeat interval -- yet the itself heartbeat would be fully controlled by the container runtime.
3. At some point a plugin might want to dispatch events.

I don't know the standard procedure in grpc/ttrpc world for making extensible backwards-compatibles APIs so that existing NRI plugin binaries would work with new runtimes and vice versa. Maybe the first request from a container runtime to a plugin could be telling what api versions the runtime can speak and ask the plugin to tell which versions it prefers. Then the communication could continue with `ConfigureRequest(config)` or `ConfigureRequestv2(config)`, for instance.

Reply:
> [low priority] This is more food-for-thought rather than a change request. But possibly something to consider for future extensibility.
> 
> At some point we might want to extend `Events` with crazy things like
> 
> * `HEARTBEAT` maybe with a couple of intervals - this would enable implementing rebalancing workloads, periodic memory arrangements etc.

Yes, agreed. But I think getting a bi-directional event/request flow between the runtime and the plugin (IOW the general ability for any plugin to make unsolicited requests to the runtime) would be a cleaner/more desirable way of doing this.

> * `PLUGIN_EVENT` that could be dispatched by a plugin and captured by another plugin and/or runtime. For instance, a plugin could have discovered new memory device being attached to the node over the fabric.
> 
> Regarding NRI-plugin backwards compatibility, these examples would mean:
> 
> 1. `ALL = 8` event is problematic, because a plugin that subscribes to every event existing now cannot know beforehand what events new NRI interface will actually serve when it starts the plugin. Do you think we could be fine with a list of explicit events to subscribe to?

Yes, we should be fine with explicit events. And that's a good point which I failed to consider altogether: backward-compatible protocol extensions. 

> 2. At some point some events could have parameters, like plugin's wish for heartbeat interval -- yet the itself heartbeat would be fully controlled by the container runtime.
> 3. At some point a plugin might want to dispatch events.
> 
> I don't know the standard procedure in grpc/ttrpc world for making extensible backwards-compatibles APIs so that existing NRI plugin binaries would work with new runtimes and vice versa. Maybe the first request from a container runtime to a plugin could be telling what api versions the runtime can speak and ask the plugin to tell which versions it prefers. Then the communication could continue with `ConfigureRequest(config)` or `ConfigureRequestv2(config)`, for instance.

Yes, good point again. We should consider more carefully what backward compatible extensibility would require/mean in practice.